nodejs/wrapper.go

```diff
@@ -0,0 +1,299 @@
+				return
+			}
+		}
+	}
+}
+
+func (nc *nodeContext) addNodeArg(nodeArg string) {
+	// find the script location and insert the provided argument
+	if len(nc.args) == 0 {
+		nc.args = []string{nodeArg}
+		return
+	}
+	for i, arg := range nc.args {
+		if len(arg) > 0 && arg[0] != '-' {
+			nc.args = append(nc.args, "")
+			copy(nc.args[i+1:], nc.args[i:])
+			nc.args[i] = nodeArg
+			logrus.Debugf("added node arg: %q", nc.args)
+			return
+		}
+	}
+	nc.args = append(nc.args, nodeArg)
+}
+
+// exec runs the command, and returns an error should one occur.
+func (nc *nodeContext) exec(in io.Reader, out, err io.Writer) error {
+	logrus.Debugf("exec: %s %v (env: %v)", nc.program, nc.args, nc.env)
+	cmd := exec.Command(nc.program, nc.args...)
+	cmd.Env = envFromMap(nc.env)
+	cmd.Stdin = in
+	cmd.Stdout = out
+	cmd.Stderr = err
+	return cmd.Run()
+}
+
+// findScript returns the path to the node script that will be executed.
+// Returns an empty string if no script was found.
+func findScript(args []string) string {
+	// a bit of a hack, but all node options are of the form `--arg=option`
+	for _, arg := range args {
+		if len(arg) > 0 && arg[0] != '-' {
+			return arg
+		}
+	}
+	return ""
+}
+
+// isApplicationScript return true if the script appears to be an application
+// script, or false if a library (node_modules) script or `npm` (special case).
+func isApplicationScript(path string) bool {
+	// We could consider checking if the parent's base name is `bin`?
+	return !strings.HasPrefix(path, "node_modules/") && !strings.Contains(path, "/node_modules/") &&
+		!strings.HasSuffix(path, "/bin/npm")
+}
+
+// envToMap turns a set of VAR=VALUE strings to a map.
+func envToMap(entries []string) map[string]string {
+	if len(entries) == 0 {
+		return nil
+	}
+	m := make(map[string]string)
+	for _, entry := range entries {
+		kv := strings.SplitN(entry, "=", 2)
+		m[kv[0]] = kv[1]
+	}
+	return m
+}
+
+// envToMap turns a map of variable:value pairs into a set of VAR=VALUE strings.
+func envFromMap(env map[string]string) []string {
+	var m []string
+	for k, v := range env {
+		m = append(m, k+"="+v)
+	}
+	return m
+}
+
+// stripInspectArg searches and removes all node `--inspect` style arguments, returning the
+// altered arguments and the inspect argument.
+func stripInspectArg(args []string) (newArgs []string, inspectArg string) {
+	// inspect directives are always a single argument: `node --inspect 9226` causes node to load 9226 as a file
+	newArgs = nil
+	inspectArg = "" // default case: no inspect arg found
+
+	for i, arg := range args {
+		if strings.HasPrefix(arg, "--inspect") {
+			// todo: we should coalesce --inspect-port=xxx
+			inspectArg = arg
+			continue
+		}
+
+		// if at end of node options, copy remaining arguments
+		// "--" marks end of node options
+		if arg == "--" || len(arg) == 0 || arg[0] != '-' {
+			newArgs = append(newArgs, args[i:]...)
+			break
+		}
+		newArgs = append(newArgs, arg)
+	}
+	return
```

Comment:
I would rather `return newArgs, inspectArg` here that use named return values.

Reply:
Ok — I think I initially had a return condition inside, so the named return args simpified things.  But that's not the case now.